packages/core/src/crypto/Jwk.ts

```diff
@@ -0,0 +1,84 @@
+import type { Ed25519Jwk, Jwk, P256Jwk, P384Jwk, P521Jwk, X25519Jwk } from './JwkTypes'
+import type { Key } from './Key'
+
+import { TypedArrayEncoder, Buffer } from '../utils'
+
+import { compress, expand } from './EcCompression'
+import {
+  isBls12381g1Jwk,
+  isBls12381g2Jwk,
+  jwkCurveToKeyTypeMapping,
+  keyTypeToJwkCurveMapping,
+  isEd25519Jwk,
+  isX25519Jwk,
+  isP256Jwk,
+  isP384Jwk,
+  isP521Jwk,
+} from './JwkTypes'
+import { KeyType } from './KeyType'
+
+export function getKeyDataFromJwk(jwk: Jwk): { keyType: KeyType; publicKey: Uint8Array } {
+  // ed25519, x25519
+  if (isEd25519Jwk(jwk) || isX25519Jwk(jwk)) {
+    return {
+      publicKey: TypedArrayEncoder.fromBase64(jwk.x),
+      keyType: jwkCurveToKeyTypeMapping[jwk.crv],
+    }
+  }
+
+  // bls12381g1, bls12381g2
+  if (isBls12381g1Jwk(jwk) || isBls12381g2Jwk(jwk)) {
+    return {
+      publicKey: TypedArrayEncoder.fromBase64(jwk.x),
+      keyType: jwkCurveToKeyTypeMapping[jwk.crv],
+    }
+  }
+
+  // p-256, p-384, p-521
+  if (isP256Jwk(jwk) || isP384Jwk(jwk) || isP521Jwk(jwk)) {
+    // TODO: do we want to use the compressed key in the Key instance?
+    const publicKeyBuffer = Buffer.concat([TypedArrayEncoder.fromBase64(jwk.x), TypedArrayEncoder.fromBase64(jwk.y)])
+    const compressedPublicKey = compress(publicKeyBuffer)
+
+    return {
+      publicKey: compressedPublicKey,
+      keyType: jwkCurveToKeyTypeMapping[jwk.crv],
+    }
+  }
+
+  throw new Error(`Unsupported JWK kty '${jwk.kty}' and crv '${jwk.crv}'`)
+}
+
+export function getJwkFromKey(key: Key): Jwk {
+  if (key.keyType === KeyType.Ed25519 || key.keyType === KeyType.X25519) {
+    return {
+      kty: 'OKP',
+      crv: keyTypeToJwkCurveMapping[key.keyType],
+      x: TypedArrayEncoder.toBase64URL(key.publicKey),
+    } satisfies Ed25519Jwk | X25519Jwk
+  }
+
+  if (key.keyType === KeyType.Bls12381g1 || key.keyType === KeyType.Bls12381g2) {
+    return {
+      kty: 'EC',
+      crv: keyTypeToJwkCurveMapping[key.keyType],
+      x: TypedArrayEncoder.toBase64URL(key.publicKey),
+    }
+  }
+
+  if (key.keyType === KeyType.P256 || key.keyType === KeyType.P384 || key.keyType === KeyType.P521) {
```

Comment:
Not really right? We could just always output the compressed variant, but support both variants the other way around

Reply:
The compressed variant, using only the X coordinate for the representation needs a prefix in the jwk of 0x03 or 0x02 I think depending on wether the y coordinate is odd or even (but that's from memory, would need a source but I am fairly sure). If we encode it without a prefix it will be interpreted incorrectly by the receiver.